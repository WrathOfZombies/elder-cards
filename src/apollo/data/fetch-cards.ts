import { reduce } from "lodash";

import { ElderCard, ElderCardPage } from "../schema";
import {
  IElderScrollsCard,
  IElderScrollsCardPage,
} from "./elder-scrolls-api.interface";

const DEFAULT_PAGE_SIZE = 20;

const BASE_URL = "https://api.elderscrollslegends.io/v1";

/**
 * Retrieves cards from the ElderScrolls API in a batch of specified
 * amount of pageSize or 20 by default. It also supports filtering based on the
 * provided search param.
 * @param {string} search Any filter that needs to be applied on the name
 * @param {string} page The current page, set this to undefined to start from the beginning
 * @param {number} pageSize The size of the current page
 * @returns {Promsie<ElderCardPage>} A promise containing an {@see ElderCardPage}
 */
export const fetchCards = async (
  search?: string,
  page?: string,
  pageSize = DEFAULT_PAGE_SIZE
): Promise<ElderCardPage> => {
  let url = `${BASE_URL}/cards?page=1&pageSize=${pageSize}`;
  url = search ? `${url}&name=${search}` : url;

  const res = await fetch(page ?? url);
  const json = await res.json();
  const data = convertToElderCardsPage(json);
  return data;
};

/**
 * Converts an array of {@see IElderScrollsCard} to the GQL normalized variant {@see ElderCard}
 * @param {IElderScrollsCard} cards The original set of cards returned from the API.
 * @returns {Array<ElderCard>} An array containing the converted card {@see ElderCard}
 */
const convertToElderCard = (
  cards: IElderScrollsCard[] | undefined
): ElderCard[] => {
  if (!cards) {
    return [];
  }

  cards = Array.isArray(cards) ? cards : [cards];
  return reduce<IElderScrollsCard, ElderCard[]>(
    cards,
    (acc, card) => {
      if (!card) {
        return acc;
      }

      try {
        const elderCard: ElderCard = {
          __typename: "ElderCard",
          id: card.id,
          name: card.name,
          rarity: card.rarity,
          type: card.type,
          subtypes: card.subtypes ?? [],
          setname: card.set?.name ?? "",
          text: card.text ?? "",
          attributes: card.attributes ?? [],
          keywords: card.keywords ?? [],
          imageUrl: card.imageUrl,
        };

        return [...acc, elderCard];
      } catch (error) {
        console.error("Failure while converting card", card);
        return acc;
      }
    },
    []
  );
};

/**
 * Converts a json containing {@see IElderScrollsCardPage} to {@see ElderCardPage}
 * @param {IElderSCrollsCardPage} json The current page returned from the API
 * @returns The GQL normalizaed variant of {@see IElderScrollsCardPage}
 */
const convertToElderCardsPage = (
  json: IElderScrollsCardPage | undefined
): ElderCardPage => {
  if (!json) {
    return {
      __typename: "ElderCardPage",
      cards: [],
      totalCount: 0,
      pageSize: 20,
      hasMore: false,
    };
  }

  const cards = convertToElderCard(json.cards);
  return {
    __typename: "ElderCardPage",
    cards,
    nextPage: json._links?.next ?? "",
    totalCount: json._totalCount ?? 0,
    pageSize: json._pageSize ?? DEFAULT_PAGE_SIZE,
    hasMore: !!json._links?.next,
  };
};
