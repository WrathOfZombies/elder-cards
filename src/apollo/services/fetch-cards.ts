import { ElderCard, ElderCardPage } from "apollo/schema";
import { reduce } from "lodash";

const DEFAULT_PAGE_SIZE = 20;
const BASE_URL = "https://api.elderscrollslegends.io/v1";

const convertToElderCard = (cards: any): ElderCard[] => {
  if (!cards) {
    return [];
  }

  cards = Array.isArray(cards) ? cards : [cards];
  return reduce<any, ElderCard[]>(
    cards,
    (acc, card) => {
      if (!card) {
        return acc;
      }

      try {
        const elderCard: ElderCard = {
          __typename: "ElderCard",
          id: card.id,
          name: card.name,
          rarity: card.rarity,
          type: card.type,
          subtype: card.subtype,
          setname: card.set?.name ?? "",
          text: card.text,
          attributes: card.attributes,
          keywords: card.keywords,
          imageUrl: card.imageUrl,
        };

        return [...acc, elderCard];
      } catch (error) {
        console.error("Failure while converting card", card);
        return acc;
      }
    },
    []
  );
};

const convertToElderCardsPage = (json: any): ElderCardPage => {
  if (!json) {
    return {
      __typename: "ElderCardPage",
      cards: [],
      totalCount: 0,
      pageSize: 20,
    };
  }

  const cards = convertToElderCard(json.cards);
  return {
    __typename: "ElderCardPage",
    cards,
    nextPage: json.links?.next,
    totalCount: json._totalCount ?? 0,
    pageSize: json._pageSize ?? DEFAULT_PAGE_SIZE,
  };
};

export const fetchCards = async (page?: string): Promise<ElderCardPage> => {
  const res = await fetch(
    page ?? `${BASE_URL}/cards?page=0&pageSize=${DEFAULT_PAGE_SIZE}`
  );
  const json = await res.json();
  return convertToElderCardsPage(json);
};
