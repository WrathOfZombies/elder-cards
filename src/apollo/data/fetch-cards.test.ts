import { fetchCards } from "./fetch-cards";

describe("Testing fetchCards", () => {
  let mockFetch: jest.Mock;

  beforeEach(() => {
    mockFetch = jest.fn();
    mockFetch.mockResolvedValue({
      json: jest.fn().mockResolvedValue({
        cards: [
          {
            name: "Cave Bear",
            rarity: "Common",
            type: "Creature",
            subtypes: ["Beast"],
            cost: 5,
            power: 5,
            health: 6,
            set: {
              id: "hos",
              name: "Heroes of Skyrim",
              _self: "https://api.elderscrollslegends.io/v1/sets/hos",
            },
            collectible: true,
            soulSummon: 50,
            soulTrap: 5,
            attributes: ["Endurance"],
            unique: false,
            imageUrl: "https://images.elderscrollslegends.io/hos/cave_bear.png",
            id: "4ce2dcfb9559e5bfee37bfb72bac55d45c95e7a8",
          },
        ],
        _links: {
          next:
            "https://api.elderscrollslegends.io/v1/cards?page=3&pageSize=20",
          prev: "https://api.elderscrollslegends.io/v1/cards?pageSize=20",
        },
        _pageSize: 20,
        _totalCount: 1212,
      }),
    });

    global.fetch = mockFetch;
  });

  test("should fetch some cards", async () => {
    const result = await fetchCards();
    expect(mockFetch).toHaveBeenCalledWith(
      "https://api.elderscrollslegends.io/v1/cards?page=1&pageSize=20"
    );
    expect(result).toMatchSnapshot();
  });

  test("should fetch some cards and search", async () => {
    await fetchCards("Alduin");
    expect(mockFetch).toHaveBeenCalledWith(
      "https://api.elderscrollslegends.io/v1/cards?page=1&pageSize=20&name=Alduin"
    );
  });

  test("should not throw if no data is returned", async () => {
    mockFetch.mockResolvedValue({
      json: jest.fn().mockResolvedValue(undefined),
    });

    const result = await fetchCards("Alduin");
    expect(result).toMatchSnapshot();
  });

  test("should not throw if cards are empty", async () => {
    mockFetch.mockResolvedValue({
      json: jest.fn().mockResolvedValue({
        cards: [],
        _links: {
          next:
            "https://api.elderscrollslegends.io/v1/cards?page=3&pageSize=20",
          prev: "https://api.elderscrollslegends.io/v1/cards?pageSize=20",
        },
        _pageSize: 20,
        _totalCount: 1212,
      }),
    });

    const result = await fetchCards("Alduin");
    expect(result).toMatchSnapshot();
  });

  test("should not throw if some properties are missing", async () => {
    mockFetch.mockResolvedValue({
      json: jest.fn().mockResolvedValue({
        cards: [
          {
            name: "Cave Bear",
            rarity: "Common",
            type: "Creature",
            cost: 5,
            power: 5,
            health: 6,

            collectible: true,
            soulSummon: 50,
            soulTrap: 5,
            text: "some content",
            unique: false,
            imageUrl: "https://images.elderscrollslegends.io/hos/cave_bear.png",
            id: "4ce2dcfb9559e5bfee37bfb72bac55d45c95e7a8",
          },
        ],
        _links: {
          next:
            "https://api.elderscrollslegends.io/v1/cards?page=3&pageSize=20",
          prev: "https://api.elderscrollslegends.io/v1/cards?pageSize=20",
        },
        _pageSize: 20,
        _totalCount: 1212,
      }),
    });

    const result = await fetchCards("Alduin");
    expect(result).toMatchSnapshot();
  });

  test("should throw if the request fails", async () => {
    mockFetch.mockRejectedValue("Something went wrong");
    expect(fetchCards()).rejects.toThrow("Something went wrong");
  });
});
