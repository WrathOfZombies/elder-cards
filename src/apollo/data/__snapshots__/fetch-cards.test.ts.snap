// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing fetchCards should fetch some cards 1`] = `
Object {
  "__typename": "ElderCardPage",
  "cards": Array [
    Object {
      "__typename": "ElderCard",
      "attributes": Array [
        "Endurance",
      ],
      "id": "4ce2dcfb9559e5bfee37bfb72bac55d45c95e7a8",
      "imageUrl": "https://images.elderscrollslegends.io/hos/cave_bear.png",
      "keywords": Array [],
      "name": "Cave Bear",
      "rarity": "Common",
      "setname": "Heroes of Skyrim",
      "subtypes": Array [
        "Beast",
      ],
      "text": "",
      "type": "Creature",
    },
  ],
  "hasMore": true,
  "nextPage": "https://api.elderscrollslegends.io/v1/cards?page=3&pageSize=20",
  "pageSize": 20,
  "totalCount": 1212,
}
`;

exports[`Testing fetchCards should not throw if cards are empty 1`] = `
Object {
  "__typename": "ElderCardPage",
  "cards": Array [],
  "hasMore": true,
  "nextPage": "https://api.elderscrollslegends.io/v1/cards?page=3&pageSize=20",
  "pageSize": 20,
  "totalCount": 1212,
}
`;

exports[`Testing fetchCards should not throw if no data is returned 1`] = `
Object {
  "__typename": "ElderCardPage",
  "cards": Array [],
  "hasMore": false,
  "pageSize": 20,
  "totalCount": 0,
}
`;

exports[`Testing fetchCards should not throw if some properties are missing 1`] = `
Object {
  "__typename": "ElderCardPage",
  "cards": Array [
    Object {
      "__typename": "ElderCard",
      "attributes": Array [],
      "id": "4ce2dcfb9559e5bfee37bfb72bac55d45c95e7a8",
      "imageUrl": "https://images.elderscrollslegends.io/hos/cave_bear.png",
      "keywords": Array [],
      "name": "Cave Bear",
      "rarity": "Common",
      "setname": "",
      "subtypes": Array [],
      "text": "some content",
      "type": "Creature",
    },
  ],
  "hasMore": true,
  "nextPage": "https://api.elderscrollslegends.io/v1/cards?page=3&pageSize=20",
  "pageSize": 20,
  "totalCount": 1212,
}
`;
